metadata_writer_nne.c

```diff
@@ -94,132 +309,725 @@ static uint8_t md_nne_handle_gps_event(struct md_writer_nne *mwn,
+                      "imsi_mccmnc=%d, "
+                      "nw_mccmnc=%d, "
+                      "cid=%d, "
+                      "rscp=%d, "
+                      "lte_rsrp=%d, "
+                      "lac=%d, "
+                      "rssi=%d, "
+                      "ecio=%d, "
+                      "lte_rssi=%d, "
+                      "lte_rsrq=%d, "
+                      "device_mode=%d, "
+                      "device_submode=%d\n",
+                      iface_event_type[mie->event_param],
+                      mie->ip_addr,
+                      mie->ifname,
+                      mie->imsi_mccmnc,
+                      mie->nw_mccmnc,
+                      mie->cid,
+                      mie->rscp,
+                      mie->lte_rsrp,
+                      mie->lac,
+                      mie->rssi,
+                      mie->ecio,
+                      mie->lte_rssi,
+                      mie->lte_rsrq,
+                      mie->device_mode,
+                      mie->device_submode);
+
+
+    struct nne_modem *modem = NULL;
+    int i;
+
+    modem = md_nne_get_modem(&(mwn->modem_list), mie->imsi_mccmnc);
+
+    if (modem == NULL)
+    {
+        modem = malloc(sizeof(struct nne_modem));
+        LIST_INSERT_HEAD(&(mwn->modem_list), modem, entries);
+        modem->mccmnc = mie->imsi_mccmnc;
+        modem->tstamp = mie->tstamp;
+
+        for(i = 0; i <= NNE_IDX_MAX; i++)
+        {
+            modem->metadata[i].tstamp = mie->tstamp;
+            modem->metadata[i].value.type = NNE_TYPE_NULL;
+        }
+
+        META_PRINT_SYSLOG(mwn->parent, LOG_ERR, "NNE writer: mccmnc %d: Added new  modem\n", mie->imsi_mccmnc);
+
+        // generate usbmodem UP event
+        struct nne_message msg;
+        msg.type = NNE_MESSAGE_TYPE_EVENT;
+        msg.tstamp = mie->tstamp;
+        msg.node = mwn->node_id;
+        msg.mccmnc = mie->imsi_mccmnc;
+        msg.key = "usbmodem";
+        msg.value = nne_value_init_str("UP");
+        msg.extra = NULL;
+        msg.source = NNE_MESSAGE_SOURCE_REPORT;
+        msg.delta = 0;
+
+        md_nne_send_message(mwn, &msg);
+
+        // Initiate cache with current values
+        for (i = 0; i < NNE_METADATA_DESCR_LEN; i++) {
+            md_nne_process_iface_event(mwn, &(NNE_METADATA_DESCR[i]), modem, mie, NNE_MESSAGE_SOURCE_QUERY);
+        }
+    }
+
+    META_PRINT_SYSLOG(mwn->parent, LOG_ERR, "NNE writer: mccmnc %d: modem %d, sec=%ld\n",
+                      mie->imsi_mccmnc, modem->mccmnc, mie->tstamp % 60);
+
+    md_nne_generate_bins1min(mwn, mie->tstamp);
+
+    modem->tstamp = mie->tstamp;
+
+    // Process metadata; only related to this iface event
+    for (i = 0; i < NNE_METADATA_DESCR_LEN; i++) {
+        if (mie->event_param == NNE_METADATA_DESCR[i].event) {
+            md_nne_process_iface_event(mwn, &(NNE_METADATA_DESCR[i]), modem, mie, NNE_MESSAGE_SOURCE_REPORT);
+        }
+    }
+
+    md_nne_check_removed_modems(mwn, mie->tstamp);
```

Comment:
Do you really want to do this for every event? Maybe it would be better to run it from a timer (which I see you already do).